Add option for secure IRC connections

atari2600tim opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
Now it can connect to IRC servers using Transport Layer Security.  `setIrcServer` includes optional boolean argument.  `getIrcServer` returns an extra boolean.  New checkbox on preferences.
#### Motivation for adding to Mudlet
Allows secure connection to IRC servers.
#### Other info (issues closed, discussion etc)
Closes #5345 
#### Release post highlight
<!--
Use this space if you wish to write a short statement or example for inclusion
in the release post for the next release. 
-->
IRC client can connect to secure servers

atari2600tim committed:
initial secure IRC

atari2600tim committed:
Merge branch 'Mudlet:development' into atari-secureIRC

atari2600tim committed:
Update TLuaInterpreter.cpp

atari2600tim committed:
Update TLuaInterpreter.cpp

atari2600tim committed:
Merge branch 'Mudlet:development' into atari-secureIRC

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -10859,17 +10859,21 @@ int TLuaInterpreter::getIrcServer(lua_State* L)
     Host* pHost = &getHostFromLua(L);
     QString hname;
     int hport;
+    bool hsecure;
```
```suggestion
    bool hsecure = false;
```
Though not modified in this PR I would recommand also initialising the previous variable:
```diff
-    int hport;
+    int hport = 0;
```

SlySven commented on src/dlgIRC.cpp:
```diff
@@ -753,6 +758,18 @@ int dlgIRC::readIrcHostPort(Host* pH)
+    } else {
+        secure = false;
+    }
+    return secure;
```
There is no need for the `bool secure` intermediate - the return value can be more concisely generated by:
```suggestion
    return secureStr.toInt();
```

SlySven commented on src/dlgIRC.cpp:
```diff
@@ -823,6 +840,11 @@ QPair<bool, QString> dlgIRC::writeIrcHostPort(Host* pH, int port)
 
+QPair<bool, QString> dlgIRC::writeIrcHostSecure(Host* pH, bool secure)
+{
+    return pH->writeProfileData(dlgIRC::HostSecureCfgItem, QString::number(secure));
```
TBH As this is a `bool` value I would *write* the strings `"true"` or `"false"` and use something like the following to process the *read* value (`text`): `text.contains(QLatin1String("true"), Qt::CaseInsensitive)` so that there is no need to decode numeric values if hand editing the setting outside of Mudlet...

SlySven commented on src/dlgIRC.h:
```diff
@@ -75,6 +79,7 @@ class dlgIRC : public QMainWindow, public Ui::irc
     QPair<bool, QString> sendMsg(const QString& target, const QString& message);
     QString getHostName() { return mHostName; }
     int getHostPort() { return mHostPort; }
+    bool getHostSecure() { return mHostSecure; }
```
All six of these *getters* would benefit from this change (insert `const` between the closing `)` and the opening `{`....

SlySven commented on src/ui/profile_preferences.ui:
```diff
@@ -2636,6 +2636,13 @@ you can use it but there could be issues with aligning columns of text</string>
             </property>
            </widget>
           </item>
+          <item row="3" column="0">
```
As this is a checkbox which does not have a separate label it is best to make it span the same space as one of the other label and control pairs (with a `colspan` attribute) and given that there is not anything on `row="2"` it does not need to be `row="3"`: 
```suggestion
          <item row="2" column="0" colspan="2">
```

SlySven commented on src/ui/profile_preferences.ui:
```diff
@@ -3729,6 +3736,7 @@ you can use it but there could be issues with aligning columns of text</string>
   <tabstop>ircNick</tabstop>
   <tabstop>ircHostPort</tabstop>
   <tabstop>ircChannels</tabstop>
+  <tabstop>ircHostSecure</tabstop>
```
As the tab order moves down the left column before jumping back to the start of the right column of controls in this block you need to reorder things so the tab order remains in that manner with the extra new control:
```suggestion
  <tabstop>ircHostSecure</tabstop>
  <tabstop>ircHostPort</tabstop>
  <tabstop>ircChannels</tabstop>
```

SlySven reviewed (commented):
Some changes to consider!

atari2600tim committed:
Merge branch 'Mudlet:development' into atari-secureIRC

atari2600tim committed:
applied suuggestions

atari2600tim committed:
missed one

this part made it where if you changed settings while connected then it would not properly connect again

atari2600tim commented on src/dlgIRC.cpp:
```diff
@@ -823,6 +834,11 @@ QPair<bool, QString> dlgIRC::writeIrcHostPort(Host* pH, int port)
 
+QPair<bool, QString> dlgIRC::writeIrcHostSecure(Host* pH, bool secure)
+{
+    return pH->writeProfileData(dlgIRC::HostSecureCfgItem, (secure ? QLatin1String("true") : QLatin1String("false")));
```
This works great as of now.  But when it saves `false` to file, it apparently throws a newline in front.  Still functional but confusing to me.  Maybe writeProfileData does something I'm not expecting.
```Alpha@Alpha-F5BZ322 MSYS /c/Users/Alpha/.config/mudlet/profiles/zzz-three
$ cat irc_secure
true
Alpha@Alpha-F5BZ322 MSYS /c/Users/Alpha/.config/mudlet/profiles/zzz-three
$ cat irc_secure

false
```

vadi2 replied:
@SlySven ?

atari2600tim committed:
Merge branch 'Mudlet:development' into atari-secureIRC

atari2600tim commented:
Should I switch the default connection from `irc.libera.chat 6667` to `irc.libera.chat +6697` also?  I don't want players to go in there and set their own server and have it automatically be set to secure and maybe confuse them.  Although it is right there and obvious enough I guess.

Otherwise SlySven caught everything that I was planning to come back to and more, very much appreciated.  I'll switch it out of draft mode.

vadi2 commented on src/ui/profile_preferences.ui:
```diff
@@ -2636,6 +2636,13 @@ you can use it but there could be issues with aligning columns of text</string>
+          <item row="2" column="0" colspan="2">
+           <widget class="QCheckBox" name="ircHostSecure">
+            <property name="text">
+             <string>TLS/SSL secure connection</string>
```
```suggestion
             <string>Use a secure connection</string>
```


+ tooltip saying it'll use TLS/SSL for the technically-minded folk who'd like more details

vadi2 reviewed (commented):
Works! Though I was confused at first because I just ticked the option, didn't realise I have to change the port as well, and the connection wasn't working. Similar problem that the connection dialog would be having...

Agree on changing the default to be more secure for new profiles only. We don't want to break existing profiles by changing them to the secure option without having updated the port...

atari2600tim committed:
add tooltip mentioning port number suggestion

Communi desktop client automatically checks the box if you type in 6697, 7000, or 7070.
https://github.com/communi/communi-desktop/blob/master/src/app/connectpage.cpp#L45

RFC 7194 just mentions 6697
